cloud/aws/aws_s3.cc

```diff
@@ -818,14 +819,80 @@ Status S3StorageProvider::CopyCloudObject(const std::string& bucket_name_src,
   return st;
 }
 
+namespace {
+class WritableFileStreamBuf : public std::streambuf {
```

Comment:
Can you please move these classes into the cloud area (out from AWS)?  Similar functionality would be useful for Azure

Reply:
It doesn't look like Azure's blob client uses iostreams for file download customization: https://github.com/Azure/azure-sdk-for-cpp/blob/azure-core_1.0.0-beta.1/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_client.hpp (thankfully)